### ISSUE TITLE
Developer Tools: add mechanism for adding script to page directly as well as via URL

### ISSUE THREAD

**Issue by CrispinF** [cloudscribe/cloudscribe.SimpleContent#482]
See discussion here https://www.cloudscribe.com/forum/thread/how-use-datatables-with-page-content-without-deployment-to-server

**Comment by markidox** [comment]
Complete

**Comment by CrispinF** [comment]
@markidox this has not actually been done, and I think needs discussion as a feature request. Re-opening.

**Comment by JimKerslake** [comment]
The main issue with doing this is that we are increasingly being pressured by customer pen test reports to implement Content Security Policies which prohibit the "Unsafe-Inline" execution of inline scripts.  (The word Unsafe says it all really...)

We've been through the majority of the cs core repo and stripped out as many of these bits of inline JS wherever possible (various stray "onClick" references on buttons for example) - we didn't get them all yet, but the aspiration would be to make cloudscribe free of any such reliance on inline scripts, so we can keep the pen testers happy.

As specified/implemented at the moment this feature would be incompatible with that - needs thought about whether there might be a way of implementing this that circumvents the 'Unsafe'ness issue.

**Comment by CrispinF** [comment]
@JimKerslake couldn't this script content be brought securely into the page with a handler?

**Comment by JimKerslake** [comment]
Yes we'd need to do something like that - can't just 'HTML.Raw' the JS content from the database straight into an inline page <script> tag.

**Comment by markidox** [comment]
I've updated this to use a tag helper. The users entered JS is then rendered to the page under a tag and then an a JS file is run which picks up that JS and dynamically inserts it into the page in a <script> tag.

Whilst this may seem like it would violate CSP it doesn't as CSP cares not about how a script is rendered in the DOM but on how the script was/is created in the first place. Because we are using a trusted script to get the JS into the page the CSP doesn't count it as being 'inline', despite it appearing inline if you were to inspect the page.

**Comment by JimKerslake** [comment]
In practical testing, for me, this did still fail to execute the script when a NWebSec CSP was in place that did not permit UnsafeInline() 

![Image](https://github.com/user-attachments/assets/96f9d2f8-752e-4a59-b6d0-33753cee5b3e)

![Image](https://github.com/user-attachments/assets/d08c604f-d60b-4a18-bf3f-b14c8b9518ba)

Maybe that's OK though... it fails cleanly and still renders that page - just no JS execution.
If we do implement such a CSP then maybe we want to prevent this kind of thing anyway... if we just state clearly on the UI that it is incompatible.

Alternatively perhaps we need to be computing, storing and delivering hashes for each dynamic script...?

**Comment by JimKerslake** [comment]
I think I can see a route to doing this without triggering the CSP police...  will make a second branch and have a tweak.

**Comment by JimKerslake** [comment]
**Important note** - I just realised that when we have a CSP that excludes UnsafeInline() in the script sources, the entire ckEditor component won't work at all, for editing any page content.

This is most likely because the ckeditor.js script itself dynamically generates its own inline <script> blocks.

This same problem does not seem (at first glance) to apply to using the Markdown editor (appending ?type=markdown) as documented here:  https://www.cloudscribe.com/using-markdown

This aspect needs testing, with regard to transitioning to any alternative content editor.  We should still aspire to getting rid of any need to rely upon 'unsafe' inline script execution.

**Comment by JimKerslake** [comment]
There are still significant usability and buggy UI issues in this - needs more work and a bit of discussion.
I'm in the middle of working on it, so will pick these up.  Missing migrations for SQLite...?!

**Comment by JimKerslake** [comment]
Obviously we've added a new Script field to cs_Page and I can see it does no harm to add one to cs_Post also in case we ever wanted to enable the same thing for blog posts (which seems very unlikely and perhaps inadvisable - but the field will do no harm).

I'm struggling to see why we also need a Script field in cs_PageResource.   I can't see we're needing that at all.  
Will re-work migrations to get rid of it again.

**Comment by JimKerslake** [comment]
Documentation re-write:  https://www.cloudscribe.com/adding-javascript-and-css-to-pages
